Components: Update ExternalLink to support onClick handler

Initsogar opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Fixes #45212

The idea is to make `<ExternalLink>` component to handle onClick events again, based on the issue #45212. Also it includes some unit testing for this component.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
`onClick` handler on `<ExternalLink>` component is currently being used in [multiple instances in Jetpack monorepo](https://github.com/Automattic/jetpack/search?q=ExternalLink+onClick) to record tracking events, and we consider it is necessary to include custom logic on that handler.

This stopped working after a change implemented for validating anchor links in https://github.com/WordPress/gutenberg/pull/42259.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
We are creating a new function for the `onClick` handler to:
* Validate and execute `onClick` whether it is passed as a prop.
* Validate if the link is an internal anchor, so we can prevent default action to not impact what was implemented on [#42259](https://github.com/WordPress/gutenberg/pull/42259)

## Testing Instructions
You can run unit tests for this:
```npm run test:unit -- external-link/test/index.tsx```

Initsogar committed:
Update ExternalLink to support onClick handler

Initsogar committed:
Add unit tests for ExternalLink component

It includes test cases for onClick event handler and preventing default action when clicking internal anchor links.

alexstine reviewed (commented):
@Initsogar Thanks for your PR. I do like how the code is a little cleaner now and makes more sense.

One thing that will be needed, the changelog will need updating at packages/components/changelog.md. Probably call this a bug fix if it is currently not working.

ciampo commented:
@chad1008 , would you be able to review this PR ?

chad1008 commented on packages/components/src/external-link/test/index.tsx:
```diff
@@ -0,0 +1,84 @@
+			</ExternalLink>
+		);
+
+		fireEvent.click( screen.getByTestId( 'external-link' ) );
```
[Looking for the element's accessible role and name](https://testing-library.com/docs/queries/byrole/) will better simulate how a user would interact with the component. Perhaps something like:

```
const component = screen.getByRole( 'link', {
			name: "I'm an anchor link! (opens in a new tab)",
		} );
```


Also along the same lines of[ testing user interaction, `userEvent` would be preferable to `fireEvent`](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/testing-overview.md#user-interactions). (examples: [ComboboxControl](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/combobox-control/test/index.js) or [ColorPalette](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/color-palette/test/index.js) tests)

chad1008 commented on packages/components/src/external-link/test/index.tsx:
```diff
@@ -0,0 +1,84 @@
+		const component = screen.getByTestId( 'external-link' );
+		const clickEvent = createEvent.click( component );
+
+		fireEvent( component, clickEvent );
```
same as above re: `userEvent` vs `fireEvent`

chad1008 commented on packages/components/src/external-link/test/index.tsx:
```diff
@@ -0,0 +1,84 @@
+		const component = screen.getByTestId( 'external-link' );
+		const clickEvent = createEvent.click( component );
+
+		fireEvent( component, clickEvent );
```
same as above re: `userEvent` vs `fireEvent`

chad1008 commented on packages/components/src/external-link/test/index.tsx:
```diff
@@ -0,0 +1,84 @@
+		const component = screen.getByTestId( 'external-link' );
+		const clickEvent = createEvent.click( component );
+
+		fireEvent( component, clickEvent );
```
same as above re: `userEvent` vs `fireEvent`

chad1008 commented on packages/components/src/external-link/test/index.tsx:
```diff
@@ -0,0 +1,84 @@
+
+		fireEvent.click( screen.getByTestId( 'external-link' ) );
+
+		expect( doSomething ).toHaveBeenCalled();
```
I'd recommend `toHaveBeenCalledTimes(1)`. The more specific check can avoid some subtle bugs from creeping in.

chad1008 commented on packages/components/src/external-link/test/index.tsx:
```diff
@@ -0,0 +1,84 @@
+
+		fireEvent( component, clickEvent );
+
+		expect( doSomething ).toHaveBeenCalled();
```
same as above re `toHaveBeenCalledTimes()` vs `toHaveBeenCalled()`

chad1008 commented on packages/components/src/external-link/test/index.tsx:
```diff
@@ -0,0 +1,84 @@
+			</ExternalLink>
+		);
+
+		const component = screen.getByTestId( 'external-link' );
```
same as above re: roles vs `getByTestId()`

chad1008 commented on packages/components/src/external-link/test/index.tsx:
```diff
@@ -0,0 +1,84 @@
+			</ExternalLink>
+		);
+
+		const component = screen.getByTestId( 'external-link' );
```
same as above re: roles vs `getByTestId()`

chad1008 commented on packages/components/src/external-link/test/index.tsx:
```diff
@@ -0,0 +1,84 @@
+			</ExternalLink>
+		);
+
+		const component = screen.getByTestId( 'external-link' );
```
same as above re: roles vs `getByTestId()`

chad1008 commented on packages/components/src/external-link/test/index.tsx:
```diff
@@ -0,0 +1,84 @@
+
+		fireEvent( component, clickEvent );
+
+		expect( clickEvent.defaultPrevented ).toBe( true );
```
With the suggested change to userEvent, you'll need a different test of the `defaultPrevented` property. Maybe something like: 

```
const onClickMock = jest.fn();

// ...

expect( onClickMock ).toHaveBeenCalledTimes( 1 );
expect( onClickMock ).toHaveBeenLastCalledWith(
  expect.objectContaining( {
    defaultPrevented: true
  } )
);
```

Initsogar replied:
@chad1008 The idea of this specific test case is to validate it is properly preventing default action if it is an anchor link, without passing onClick prop to our component (I will update the test case name). I'm trying to find a way to accomplish this using userEvent, but don't find any yet. Do you have any idea how we could handle it?

Initsogar replied:
I think we can mock the `onClick` after rendering the link from the component and validate we are receiving the `defaultPrevented: true`. Something like:

```
const user = setupUser();

render(
	<ExternalLink href="#test">I&apos;m an anchor link!</ExternalLink>
);

const link = screen.getByRole( 'link', {
	name: "I'm an anchor link! (opens in a new tab)",
} );

const onClickMock = jest.fn();
link.onclick = onClickMock;

await user.click( link );
expect( onClickMock ).toHaveBeenCalledTimes( 1 );
expect( onClickMock ).toHaveBeenLastCalledWith(
	expect.objectContaining( { defaultPrevented: true } )
);
```

What do you think?

Initsogar replied:
I tested this approach, and it is working as expected. I will push new changes so you can review it properly :)

chad1008 replied:
Ah, yes I see your point... hard to test the behavior when no `onClick` is passed by ...passing a mock to `onClick` 😆

Your proposed solution looks like works, although modifying the event listener imperatively is an unusual pattern. Maybe we just add a brief comment to those two tests explaining that we're using that approach specifically so we can test the `defaultPrevented` without passing an `onClick` prop? It's described in the test title, but a comment closer to the actual `link.onclick` modification might help make it crystal clear for future maintainers.

Initsogar replied:
Correct! 🥲 I just added the comments!

chad1008 commented on packages/components/src/external-link/test/index.tsx:
```diff
@@ -0,0 +1,84 @@
+		fireEvent( component, clickEvent );
+
+		expect( doSomething ).toHaveBeenCalled();
+		expect( clickEvent.defaultPrevented ).toBe( true );
```
same as above re: `clickEvent.defaultPrevented` after implementing `userEffect`

chad1008 commented on packages/components/src/external-link/test/index.tsx:
```diff
@@ -0,0 +1,84 @@
+
+		fireEvent( component, clickEvent );
+
+		expect( clickEvent.defaultPrevented ).toBe( false );
```
same as above re: `clickEvent.defaultPrevented` after implementing `userEffect`

chad1008 reviewed (changes requested):
Changes to the component look good to me 🙂. I've included some inline comments on the new unit tests.

Other than that I think all this will need is a CHANGELOG entry!

Initsogar commented:
Thanks @alexstine and @chad1008, I will be working on the suggested changes!

Initsogar committed:
Test using getByRole and userEvent

Initsogar commented:
Hmm, I clicked to ask for re-review from @chad1008 and @alexstine and somehow it messed up and removed the other ones that were on the list. Sorry 😅

I just added the changelog entry and made changes on the unit test related to use getByRole and userEvent instead of getByTestId and fireEvent/createEvent. I would appreciate your review :)

Initsogar committed:
Add comments for defaultPrevented approach

Initsogar committed:
Add CHANGELOG entry

Initsogar commented:
@chad1008 I just added some comments for clarification purposes on the approach used for unit testing `defaultPrevented` on this component while no passing `onClick` prop. Also resolved the conflict with components `CHANGELOG.md`.

chad1008 commented on packages/components/src/external-link/test/index.tsx:
```diff
@@ -0,0 +1,113 @@
+			<ExternalLink
+				href="#test"
+				onClick={ onClickMock }
+				data-testid="external-link"
```
Just noticed this `data-testid` is still here, but no longer needed. Looks like there's another lingering in the next test as well. We can remove both before merging 🙂

Initsogar replied:
You are right! Thank you for that!
Just [fixed](https://github.com/WordPress/gutenberg/pull/45214/commits/0f48906784b35bce8d4ab9197ce487f97436dcb3) it :)

chad1008 reviewed (approved):
Looks great @Initsogar, thank you! I left one last inline comment (sorry for not spotting that last time), but after that I think we're good to go 🚀 🚢 

Congrats on your first core contribution, btw! 🎉

Initsogar committed:
Remove unnecessary data-testid attributes

Initsogar commented:
@chad1008 @ciampo @mirka Hey there! I pushed the latest suggestion and I think we are ready. Could you confirm and let me know what would be the next step? :) Thanks for all your support!

ciampo commented:
Hey @Initsogar , I re-ran the failing tests (it looks like they're being a flakey lately) and they passed!

I'll go ahead and merge your changes — thank you for helping the project, and congrats on your first contribution 🎉  Happy to collaborate more in the future in case you felt like opening more PRs like this one
